common/persistence/queue_v2_util.go

```diff
@@ -54,22 +56,92 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
 	return append([]byte{pageTokenPrefixByte}, b...)
 }
 
-func GetMinMessageIDForQueueV2(request *InternalReadMessagesRequest) (int64, error) {
-	// TODO: start from the ack level of the queue partition instead of the first message ID when there is no token.
-	if len(request.NextPageToken) == 0 {
-		return FirstQueueMessageID, nil
+func GetMinMessageIDToReadForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	nextPageToken []byte,
+	queue *persistencespb.Queue,
+) (int64, error) {
+	if len(nextPageToken) == 0 {
+		partition, err := GetPartitionForQueueV2(queueType, queueName, queue)
+		if err != nil {
+			return 0, err
+		}
+		return partition.MinMessageId, nil
 	}
 	var token persistencespb.ReadQueueMessagesNextPageToken
 
 	// Skip the first byte. See the comment on pageTokenPrefixByte for more details.
-	err := token.Unmarshal(request.NextPageToken[1:])
+	err := token.Unmarshal(nextPageToken[1:])
 	if err != nil {
 		return 0, fmt.Errorf(
 			"%w: %q: %v",
 			ErrInvalidReadQueueMessagesNextPageToken,
-			request.NextPageToken,
+			nextPageToken,
 			err,
 		)
 	}
 	return token.LastReadMessageId + 1, nil
 }
+
+func GetPartitionForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	queue *persistencespb.Queue,
+) (*persistencespb.QueuePartition, error) {
+	// Currently, we only have one partition for each queue. However, that might change in the future. If a queue is
+	// created with more than 1 partition by a server on a future release, and then that server is downgraded, we
+	// will need to handle this case. Since all DLQ tasks are retried infinitely, we just return an error.
+	numPartitions := len(queue.Partitions)
+	if numPartitions != 1 {
+		return nil, serviceerror.NewInternal(
+			fmt.Sprintf(
+				"queue with type %v and queueName %v has %d partitions, but this implementation only supports"+
```

Comment:
nit: It's easier to search for where error messages occurred in the code base if error logs are structured like this <some static message>: <something formatted>, so I think we should change this to be something like "queue without single partition detected: ..."

Reply:
Done